src/librustc/diagnostics.rs

```diff
@@ -1923,54 +1932,58 @@ From [RFC 246]:
 "##,
 
 E0395: r##"
-The value assigned to a constant expression must be known at compile time,
-which is not the case when comparing raw pointers. Erroneous code example:
+The value assigned to a constant scalar must be known at compile time,
+which is not the case when comparing raw pointers.
 
+
+Erroneous code example:
 ```
-static foo: i32 = 42;
-static bar: i32 = 43;
+static FOO: i32 = 42;
+static BAR: i32 = 42;
 
-static baz: bool = { (&foo as *const i32) == (&bar as *const i32) };
+static BAZ: bool = { (&FOO as *const i32) == (&BAR as *const i32) };
 // error: raw pointers cannot be compared in statics!
 ```
 
-Please check that the result of the comparison can be determined at compile time
-or isn't assigned to a constant expression. Example:
+The address assigned by the linker to FOO and BAR may or may not
+be identical, so the value of BAZ can't be determined.
+
+If you want to do the comparison, please do it at run-time.
+
+For example:
 
 ```
-static foo: i32 = 42;
-static bar: i32 = 43;
+static FOO: i32 = 42;
+static BAR: i32 = 42;
 
-let baz: bool = { (&foo as *const i32) == (&bar as *const i32) };
+let baz: bool = { (&FOO as *const i32) == (&BAR as *const i32) };
 // baz isn't a constant expression so it's ok
 ```
 "##,
 
 E0396: r##"
-The value assigned to a constant expression must be known at compile time,
-which is not the case when dereferencing raw pointers. Erroneous code
-example:
+The value behind a raw pointer can't be determined at compile-time
+(or even link-time), which means it can't be used in a constant
+expression.
 
+For example:
 ```
-const foo: i32 = 42;
-const baz: *const i32 = (&foo as *const i32);
+const REG_ADDR: *const u8 = 0x5f3759df as *const u8;
 
-const deref: i32 = *baz;
+const VALUE: u8 = unsafe { *REG_ADDR };
 // error: raw pointers cannot be dereferenced in constants
 ```
 
-To fix this error, please do not assign this value to a constant expression.
-Example:
+To fix this, you should dereference your pointer at some point in
+run-time.
```

Comment:
I am not sure how important is the "how to fix this" advice, because each issue will have its own fix.